pkg/shared/utils/containerRootfs/fs.go

```diff
@@ -0,0 +1,166 @@
+	"github.com/anchore/stereoscope/pkg/file"
+	"github.com/anchore/stereoscope/pkg/filetree/filenode"
+)
+
+const perFileReadLimit = 2 * file.GB
+
+func ToDirectory(ctx context.Context, src, dest string) error {
+	image, err := stereoscope.GetImage(ctx, src)
+	if err != nil {
+		return fmt.Errorf("unable to parse image from src %s: %w", src, err)
+	}
+	defer image.Cleanup()
+
+	err = image.SquashedTree().Walk(func(path file.Path, f filenode.FileNode) error {
+		target := filepath.Join(dest, string(path))
+
+		switch f.FileType {
+		case file.TypeDirectory:
+			if err := os.MkdirAll(target, 0755); err != nil {
+				return err
+			}
+		case file.TypeRegular:
+			indexEntry, err := image.FileCatalog.Get(*f.Reference)
+			if err != nil {
+				return fmt.Errorf("unable to get index entry for file: %w", err)
+			}
+
+			output, err := os.OpenFile(target, os.O_CREATE|os.O_RDWR, indexEntry.Mode())
+			if err != nil {
+				return fmt.Errorf("unable to open file %s: %w", target, err)
+			}
+
+			content, err := image.OpenReference(*f.Reference)
+			if err != nil {
+				return fmt.Errorf("unable to open path from squash: %w", err)
+			}
+
+			numBytes, err := io.Copy(output, io.LimitReader(content, perFileReadLimit))
+			if numBytes >= perFileReadLimit || errors.Is(err, io.EOF) {
+				return fmt.Errorf("zip read limit hit (potential decompression bomb attack)")
+			}
+			if err != nil {
+				return fmt.Errorf("unable to copy file: %w", err)
+			}
+
+			err = output.Close()
+			if err != nil {
+				// TODO log the error
+			}
+		}
+
+		return nil
+	}, nil)
+	if err != nil {
+		return fmt.Errorf("unable to walk squashed tree from container: %w", err)
+	}
+
+	return nil
+}
+
+type Rootfs interface {
+	Dir() string
+	Cleanup() error
+}
+
+func ToTempDirectory(ctx context.Context, src string) (Rootfs, error) {
+	tdrf := &TempDirRootfs{}
+
+	tmpdir, err := os.MkdirTemp("", "")
+	if err != nil {
+		return tdrf, fmt.Errorf("unable to create temp directory: %w", err)
+	}
+	tdrf.dir = tmpdir
+
+	err = ToDirectory(ctx, src, tmpdir)
+	if err != nil {
+		return tdrf, fmt.Errorf("unable output squashed image to directory: %w", err)
+	}
+
+	return tdrf, nil
+}
+
+type TempDirRootfs struct {
+	dir string
+}
+
+func (tdrf *TempDirRootfs) Dir() string {
+	return tdrf.dir
+}
+
+func (tdrf *TempDirRootfs) Cleanup() error {
+	var errs []error
+
+	if tdrf.dir != "" {
+		err := os.RemoveAll(tdrf.dir)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("unable to remove temp directory: %w", err))
+		}
+	}
+
+	return errors.Join(errs...)
+}
+
+type cachedRootfs struct {
+	complete chan struct{}
+	err      error
+	rootfs   Rootfs
+}
+
+var cacheMutex sync.RWMutex
+var cache map[string]*cachedRootfs = map[string]*cachedRootfs{}
+
+func (crfs *cachedRootfs) Dir() string {
+	return crfs.rootfs.Dir()
+}
+
+func (crfs *cachedRootfs) Cleanup() error {
+	return nil
+}
+
+func ToCachedTempDirectory(ctx context.Context, src string) (Rootfs, error) {
+	cacheMutex.Lock()
```

Comment:
why not to use `sync.Once`?

Reply:
I'm not sure sync.Once will work here, when we have two callers asking to convert the same input concurrently, we need to be able to check if the conversion is already in progress for that particular src, and if it is we need to wait, if it is not then we need to start it.

Even if we used sync.Once we need to store a "completed" channel somewhere, and we'd also need to store the instance of sync.Once somewhere so we'd still need a lock for that.

Reply:
`sync.Once` will make sure that function only called once, right?
So, if the function of conversion only run once and the path of the dir is kept on the struct (and maybe the error), other callers will just get the path (or the error) as a return.

something like this?
```
var (
	once sync.Once
)

type Rootfs interface {
	Dir() string
	Cleanup() error
}

type TempDirRootfs struct {
	dir string
}

type cachedTempDir struct {
	tdrf  *TempDirRootfs
	error error
}

func (c *cachedTempDir) ToCachedTempDirectory(ctx context.Context, src string) (Rootfs, error) {
	once.Do(func() {
		rootfs, err := ToTempDirectory(ctx, src)
		c.tdrf = rootfs
		c.error = err
	})
	return c.tdrf, c.error
}
```

Reply:
But I see that I missed the handling of the logic for different sources, so maybe `sync.Once` is not the case here.

Reply:
ToCachedTempDirectory can be called with multiple srcs i.e.

ToCachedTempDirectory(ctx, "nginx.tar")
ToCachedTempDirectory(ctx, "grype-server.tar")

In your implementation, only the first src will ever be unpacked. Also if we have two scanners calling it:

ToCachedTempDirectory(ctx, "nginx.tar")
ToCachedTempDirectory(ctx, "nginx.tar")

the second caller will get an partially unpacked directory because ToTempDirectory is still running.

Reply:
> the second caller will get an partially unpacked directory because ToTempDirectory is still running.

not sure why is that, but nm I think I get why sync.Once is not a good option here

Reply:
sync.Once() only guarantees that the passed func is called exactly once, following callers aren't blocked if that call is still running.

```
func Do() {
    once.Do(func() { time.Sleep(infinity) })
}

go func() {
    Do() <--- sleeps forever
}()

go func() {
   Do() <--- returns instantly
}()
```

Reply:
I'm not sure that is correct, based on https://pkg.go.dev/sync#Once.Do 
> Because no call to Do returns until the one call to f returns, if f causes Do to be called, it will deadlock.

https://go.dev/play/p/jqU5WQoyORm

Reply:
huh... well I've misread that part of the docs so many times, sync.Once is much more useful in that case!

Reply:
😄 
maybe we can hold `once sync.Once` per `src` - something like 
```
type TempDirRootfs struct {
	dir string
    once sync.Once
}

var cache map[string]*TempDirRootfs
```

Reply:
instead of 
```
type cachedRootfs struct {
	complete chan struct{}
	err      error
	rootfs   Rootfs
}

type Cache struct {
	mu sync.Mutex
	m  map[string]*cachedRootfs
}
```
we can use
```
type cachedRootfs struct {
	once sync.Once
	err      error
	rootfs   Rootfs
}

type Cache struct {
	mu sync.Mutex
	m  map[string]*cachedRootfs
}
```

will it not simplified the complete/done/wait logic?

so you can getOrSet the cachedRootfs under mutex, and use the `once` to run only once the `toTempDirectory()`?

Reply:
I'm not sure that it will now, especially given Erez's comment about make sure that we abort Wait() if the context is canceled. I've updated the Wait and Done functions to make this nicer.